src/mono/mono/metadata/icall.c

```diff
@@ -3436,19 +3524,87 @@ ves_icall_InternalInvoke (MonoReflectionMethodHandle method_handle, MonoObjectHa
 		} else {
 			g_assert (pcount == (m_class_get_rank (m->klass) * 2));
 			/* The arguments are lower-bound-length pairs */
-			intptr_t * const lower_bounds = (intptr_t *)g_alloca (sizeof (intptr_t) * pcount);
+			intptr_t *lower_bounds = (intptr_t *)g_alloca (sizeof (intptr_t) * pcount);
 
 			for (int i = 0; i < pcount / 2; ++i) {
-				lower_bounds [i] = *(int32_t*) ((char*)mono_span_get (params_span, MonoObject*, (i * 2)) + sizeof (MonoObject));
-				lengths [i] = *(int32_t*) ((char*)mono_span_get (params_span, MonoObject*, (i * 2) + 1) + sizeof (MonoObject));
+				lower_bounds [i] = *(int32_t*) params_byref [i * 2];
+				lengths [i] = *(int32_t*) params_byref [i * 2 + 1];
 			}
 
 			arr = mono_array_new_full_checked (m->klass, lengths, lower_bounds, error);
 			goto_if_nok (error, return_null);
 			goto exit;
 		}
 	}
-	result = mono_runtime_invoke_span_checked (m, obj, params_span, error);
+
+	int params_length = sig->param_count;
+	gpointer *pa = NULL;
+	if (params_length > 0) {
+		pa = g_newa (gpointer, params_length);
+		for (int i = 0; i < params_length; i++) {
+			MonoType *t = sig->params [i];
+			pa [i] = invoke_byrefs_extract_argument (params_byref, i, t, error);
+			goto_if_nok (error, return_null);
+		}
+	}
+	if (!strcmp (m->name, ".ctor") && m->klass != mono_defaults.string_class) {
+		gpointer o = obj;
+		if (mono_class_is_nullable (m->klass)) {
+			/* Need to create a boxed vtype instead */
+			g_assert (!obj);
+
+			if (params_length == 0) {
+				goto_if_nok (error, return_null);
+			} else {
+				result = mono_value_box_checked (m_class_get_cast_class (m->klass), pa [0], error);
+				goto exit;
+			}
+		}
+
+		if (!obj) {
+			MonoObjectHandle obj_h = mono_object_new_handle (m->klass, error);
+			goto_if_nok (error, return_null);
```

Comment:
Might be better to use the non _handle functions and MONO_HANDLE_PIN () in this function.